subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/DynamicVersionResolver.java

```diff
@@ -262,14 +264,14 @@ protected void applyTo(ResourceAwareResolveResult target, Set<String> unmatchedV
         private final ModuleComponentRepository repository;
         private final AttemptCollector attemptCollector;
         private final DependencyMetadata dependencyMetadata;
-        private final String version;
+        private final Version version;
         private boolean searchedLocally;
         private boolean searchedRemotely;
         private final DefaultBuildableModuleComponentMetaDataResolveResult result = new DefaultBuildableModuleComponentMetaDataResolveResult();
 
         public CandidateResult(DependencyMetadata dependencyMetadata, String version, ModuleComponentRepository repository, AttemptCollector attemptCollector) {
             this.dependencyMetadata = dependencyMetadata;
-            this.version = version;
+            this.version = VersionParser.INSTANCE.transform(version);
```

Comment:
One improvement we could do is to use a factory here instead of a global singleton. As a service, it would allow us to cache results, and avoid parsing again and again. It would at least reduce memory pressure.